File Renaming Plugin #POL-1109

mahsheed opened the pull request:
This WIPP plugin uses user-supplied file naming patterns to dynamically rename and save files in an image collection to a new output directory.  #POL-1109

Nicholas-Schaub commented on polus-file-renaming-plugin/src/main.py:
```diff
@@ -0,0 +1,297 @@
+    out_patts = re.findall(r"\{(\w+):([dc+]+)\}", out_pattern)
+    unique_keys = []
+    #: If input file pattern is c and output is d, list unique key
+    unique_keys = list(set([
+        inp_grp for (inp_grp,inp_rgx) in in_patts 
+        for (out_grp, out_rgx) in out_patts  
+        if inp_rgx.startswith("c") and out_rgx.startswith("d")
+        ]))
```
Is this assuming there is only one categorical variable we want to convert to numeric? What happens if we have two string variables we want to convert to numeric variables?

mahsheed replied:
It can handle multiple categorical variables to convert to numeric. I added new tests to show it processing two and three string variables we want to convert to numeric (`test_two_chan_to_digit`, `test_three_chan_to_digit`)

If we have two string variables to convert to numeric, `non_numstr_value_to_int()` produces a dictionary for each named group where the key is the captured string for the named group and the value is the index of the string's alphabetical position. For example, `non_numstr_value_to_int()` would return `{'DAPI': 0, 'GFP': 1, 'TXRED': 2}` for one capture group (ex: `{chan:c+}`), then, `non_numstr_value_to_int()` would return `{'ABC': 0, 'DEF': 1, 'GHI': 2}` for the next capture group (ex: `{alpha_chan:ccc}`). 

Then, these values get processed and incorporated into the appropriate positions when generating the output filename.

Nicholas-Schaub commented on polus-file-renaming-plugin/src/main.py:
```diff
@@ -0,0 +1,297 @@
+    inp_pattern = filePattern
+    out_pattern = outFilePattern
+    
+    """Process Inputs"""
+    #: Generate dict of named regular expression groups
+    regex_patterns = pattern_to_regex(inp_pattern)
+    #: Update file pattern to be regex-compatible, using dict of regex_patterns
+    inp = pattern_to_raw_f_string(inp_pattern, regex_patterns)
+    #: Convert out pattern to fstring to be used for str formatting
+    out_pattern_fstring = pattern_to_fstring(out_pattern)
+    #: List category names where input, output patterns are c, d)
+    category_list = replace_cat_label(inp_pattern, out_pattern)
+    #: List dict for each filename, capture groups, and corresponding matches
+    all_matches = gen_all_matches(inp, inp_files)
+    #: Iterate through matches
+    for i in range(0, len(all_matches)):
+        tmp_match = all_matches[i]
+        #: Convert numeric str dictionary values to int, if applicable
+        all_matches[i] = numstrvalue_to_int(tmp_match)
+    char_to_num = dict()
+    for c_to_d_category in category_list:
+        char_to_num[c_to_d_category] = non_numstr_value_to_int(
+            c_to_d_category, all_matches)
+    #: Convert tmp_match to integers, if applicable
+    for c_to_d_category in category_list:
+        for i in range(0, len(all_matches)):
+            if all_matches[i].get(c_to_d_category):
+                all_matches[i][c_to_d_category] = char_to_num[c_to_d_category][
+                    all_matches[i][c_to_d_category]]
+    for match in all_matches:
+        out_name = Path(outDir).resolve() / out_pattern_fstring.format(**match)
+        logger.info(f"Old file name {match['fname']} and new file name {out_name}")
+        #: Copy renamed file to output directory
```
This looks good, but please move this to a `main` function with typed inputs. We are trying to only have input parsing done under the `if __name__=='__main__` block and then pass those parsed inputs to a `main` function. The reason is so that we can call this code from another Python function if we want.

mahsheed replied:
Thank you, that makes sense. I will move the needed parts to main()

Nicholas-Schaub reviewed (changes requested):
I was a little nitpicky about things because I was reviewing this not only as a PR, but also doing some thinking about this in terms of a new filepattern syntax.

Some things I didn't include in the comments, but are more general questions:
1. What happens when `()` are in the file names? It seems like since you're using named groups, arbitrary `()` shouldn't matter much in the file names. However, I commonly see parentheses used around things like channel (i.e. `(DAPI)`), so we should add a test for it.
2. We should add in some tests to make sure that non-alphanumeric characters are handled properly. Thinks like spaces, periods, etc.
3. Instead of including actual images, it might be better to use a dictionary of stored values to test against. I did this in the `filepattern` utility for testing. This will allow you to have a larger, more diverse set of test data. The downside is that you may need to tweak your code to get the tests to work. You can see what I am suggesting here:
https://github.com/PolusAI/filepattern/tree/master/tests

I think overall this looks really good. I think it could use some additional code consolidation, but this is a really good piece of work as is.

mahsheed commented on polus-file-renaming-plugin/plugin.json:
```diff
@@ -0,0 +1,56 @@
+    "version": "0.1.13",
+    "title": "File Renaming",
+    "description": "Rename and store image collection files in a new image collection",
+    "author": "Melanie Parham (melanie.parham@axleinfo.com)",
```
Thanks for the clarification. Because I am still in the middle of being transferred to NCATS from NIAID, I do not have access to my previous NIH email. However, I can add it later after I regain access.

mahsheed commented on polus-file-renaming-plugin/src/main.py:
```diff
@@ -0,0 +1,297 @@
+    return out_pattern_fstring
+
+def replace_cat_label(inp_pattern:str, out_pattern:str)->list:
+    """Return categorical labels if inpatt is char and outpatt is digit.
```
It is not a typo, but I will update the variable names to be more descriptive.

mahsheed commented on polus-file-renaming-plugin/src/main.py:
```diff
@@ -0,0 +1,297 @@
+    out_patterns = re.findall(r"\{(\w+):([dc+]+)\}", out_pattern)
+    f_string_dict = {}
+    for key, value in out_patterns:
+        temp_value = value[:1]
+        if "+" not in value and temp_value == "c":
+            temp_value = "s"
+            f_string_dict[key] = "{" + key + ":" + str(len(value)) + temp_value + "}"
+        # Prepend "0" to field to enable 0-padding of numeric types
+        elif "+" not in value and temp_value != "c":
+            f_string_dict[key] = "{" + key + ":0" + str(len(value)) + temp_value + "}"
+        elif "+" in value and temp_value == "c":
+            temp_value = "s"
+            f_string_dict[key] = "{" + key  + ":" + temp_value + "}"
+        # Prepend "0" to field to enable 0-padding of numeric types
+        elif "+" in value and temp_value != "c":
+            f_string_dict[key] = "{" + key  + ":0" + temp_value + "}"
```
Thanks for the suggestion. I will test it out and try implementing this more simplified approach.

mahsheed commented on polus-file-renaming-plugin/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+FROM python:3.9-alpine
+COPY VERSION /
+		
+ARG EXEC_DIR="/opt/executables"
+ARG DATA_DIR="/data"
```
Thank you for pointing me in the right direction. I will update the code accordingly.

mahsheed commented on polus-file-renaming-plugin/src/main.py:
```diff
@@ -0,0 +1,297 @@
+        rgx_patterns: named group and regex value dictionary
+    """
+    #: Return reference to logger instance
+    logger = logging.getLogger("main")
```
Thank you for your suggestion. I have implemented your suggestion. However, I decided to also define it in the `gen_all_matches()` function, in order to have it log more detailed exceptions.

mahsheed commented:
Thank you @Nicholas-Schaub for reviewing my PR. I appreciated all of your suggestions.
I refactored the code based on the python template (adjusting `main()`) and worked on cleaning the code (adding more descriptive function/variable names, improved commenting). I added more test cases as well. 

Regarding your questions,

1. The plugin now works with parentheses in filenames.
2. I added tests for non-alphanumeric characters. I also enabled the plugin to handle special characters that would interfere with regex.
3. There is now `file_rename_test.json` for easier filename testing.

The new pull request with these changes is located here: https://github.com/PolusAI/polus-plugins/pull/210
